vllm/sequence.py

```diff
@@ -112,50 +142,66 @@ class SequenceData:
         prompt_token_ids: The token IDs of the prompt.
         output_token_ids: The token IDs of the output.
         cumulative_logprob: The cumulative log probability of the output.
+        tokens: array of all the tokens (prompt + output)
+    
+    NOTE: special care must be taken regarding data copy and returning `list`
+        or `np.ndarray` from this class. `get_prompt_token_ids` and 
+        `get_output_token_ids` return `list`. They are used to construct
+        request output that will be returned to the user. They need to be
+        Python lists. And they are actually quite cheap, because the function
+        returns a reference to the list that is already stored in the object.
+        `get_token_ids` returns a view of `np.ndarray`. It avoids data copy,
+        and also allows array operations to be performed on the data.
     """
 
     def __init__(
         self,
         prompt_token_ids: List[int],
         output_token_ids: Optional[List[int]] = None,
     ) -> None:
+        self.tokens = _SEQUENCE_DATA_POOL.alloc_array()
+        self.prompt_token_ids_list = prompt_token_ids
```

Comment:
Is there any opportunity to get rid of this list (and output token ids list)? This is completely duplicated to the numpy arrays and we should avoid that as possible.

Reply:
I searched the code base and seems like only batch expansion uses `get_prompt_token_ids()` and `get_output_token_ids()`, so it should be possible, as batch expansion is going to be removed by @LiuXiaoxuanPKU

Reply:
good to know.